### ISSUE TITLE
tensorflow.dll build with GPU support fails on Windows with "( was unexpected at this time"

### ISSUE THREAD

**Issue by MikhailStartsev** [tensorflow/tensorflow#39905]
**System information**
- OS Platform and Distribution: Windows 10 Enterprise 1809
- TensorFlow installed from (source or binary): source
- TensorFlow version: 2.2.0
- Python version: 3.5.6 (via Anaconda)
- Bazel version (if compiling from source): 2.0.0
- GCC/Compiler version (if compiling from source): MS VisualStudio 2019 14.25.28610
- CUDA/cuDNN version: 10.1/7.6.5.32
- GPU model and memory: NVIDIA TITAN Xp 12GB

**Describe the problem**

Building a TF2 for GPU on Windows craches with a (likely cmd-related) error:
```( was unexpected at this time.```

This happens when building //tensorflow:tensorflow.dll, but in particular for any *_gpu target (i.e. can build a smaller target from the very beginning, will lead to the same error - alternative "short" command provided below as well)

**Provide the exact sequence of commands / steps that you executed before running into the problem**

Step 0: Setting up & activating a clean virtual environment as per instructions at https://www.tensorflow.org/install/source_windows, installing & upgrading MSYS2, placing [Bazel 2.0.0](https://github.com/bazelbuild/bazel/releases/download/2.0.0/bazel-2.0.0-windows-x86_64.exe) into the appropriate folder.

Step 1: Executing the following in cmd:

```
cd C:\source
git clone https://github.com/tensorflow/tensorflow.git
cd tensorflow
git checkout v2.2.0

:: setting up paths
SET PATH=C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\bin;%PATH%
SET PATH=C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\include;%PATH%
SET PATH=C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\extras\CUPTI\lib64;%PATH%
:: (CUDA_PATH points to the same v10.1 folder)
:: (cuDNN files already copied into the respective cuda folder)
SET PATH=C:\Program Files (x86)\Bazel;%PATH%
SET PATH=C:\msys64\usr\bin;%PATH%

:: setting up Bazel-related environment variables
set BAZEL_VS=C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional
set BAZEL_VC=C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC
set BAZEL_VC_FULL_VERSION=14.25.28610
:: (not setting BAZEL_SH since added msys64 into path, but setting it also does not help)
```

Configuration script output:
```
(TF2Build) C:\source\tensorflow>configure.cmd
You have bazel 2.0.0 installed.
Please specify the location of python. [Default is C:\Users\mikhail.startsev\Venvs\TF2Build\Scripts\python.exe]:


Found possible Python library paths:
  C:\Users\mikhail.startsev\Venvs\TF2Build\lib\site-packages
Please input the desired Python library path to use.  Default is [C:\Users\mikhail.startsev\Venvs\TF2Build\lib\site-packages]

Do you wish to build TensorFlow with ROCm support? [y/N]: n
No ROCm support will be enabled for TensorFlow.

Do you wish to build TensorFlow with CUDA support? [y/N]: y
CUDA support will be enabled for TensorFlow.

Found CUDA 10.1 in:
    C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/lib/x64
    C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include
Found cuDNN 7 in:
    C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/lib/x64
    C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include


Please specify a list of comma-separated CUDA compute capabilities you want to build with.
You can find the compute capability of your device at: https://developer.nvidia.com/cuda-gpus.
Please note that each additional compute capability significantly increases your build time and binary size, and that TensorFlow only supports compute capabilities >= 3.5 [Default is: 3.5,7.0]:


Please specify optimization flags to use during compilation when bazel option "--config=opt" is specified [Default is /arch:AVX]:


Would you like to override eigen strong inline for some C++ compilation to reduce the compilation time? [Y/n]: y
Eigen strong inline overridden.

Would you like to interactively configure ./WORKSPACE for Android builds? [y/N]: n
Not configuring the WORKSPACE for Android builds.
```

What I want to build in the end:
```
bazel build --define=no_tensorflow_py_deps=true -c opt //tensorflow/tensorflow.dll
```

What give the same errors but potentially faster (any other *_gpu target will do as well):
```
bazel build --define=no_tensorflow_py_deps=true -c opt tensorflow/core/kernels/debug_ops_gpu
```

Both result in the following lines at the end of the output (for //tensorflow/tensorflow.dll, the sub-target varies with each run, but always *_gpu):

```ERROR: C:/source/tensorflow/tensorflow/core/kernels/BUILD:2046:1: C++ compilation of rule '//tensorflow/core/kernels:gather_functor_gpu' failed (Exit 255)
( was unexpected at this time.
Target //tensorflow/core/kernels:debug_ops_gpu failed to build
```

**Any other info / logs**

For a more explicit output I ran the build with these extra options: `-s --verbose_failures --jobs=1`. The terminal output of the last subcommand is below:

``` 
INFO: Analyzed target //tensorflow/core/kernels:debug_ops_gpu (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
SUBCOMMAND: # //tensorflow/core/kernels:debug_ops_gpu [action 'Compiling tensorflow/core/kernels/debug_ops_gpu.cu.cc', configuration: 0da6e0533905602ae3b4151f84e97ac897aad4120d80febd63f279c70317ba99]
cd C:/users/mikhail.startsev/_bazel_mikhail.startsev/3watukzj/execroot/org_tensorflow
  SET CUDA_TOOLKIT_PATH=C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1
    SET INCLUDE=C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.25.28610\ATLMFC\include;C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.25.28610\include;C:\Program Files (x86)\Windows Kits\NETFXSDK\4.8\include\um;C:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt;C:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\shared;C:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\um;C:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\winrt;C:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\cppwinrt
    SET LIB=C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.25.28610\ATLMFC\lib\x64;C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.25.28610\lib\x64;C:\Program Files (x86)\Windows Kits\NETFXSDK\4.8\lib\um\x64;C:\Program Files (x86)\Windows Kits\10\lib\10.0.18362.0\ucrt\x64;C:\Program Files (x86)\Windows Kits\10\lib\10.0.18362.0\um\x64;
    SET PATH=C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.25.28610\bin\HostX64\x64;C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\Common7\IDE\VC\VCPackages;C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\Common7\IDE\CommonExtensions\Microsoft\TestWindow;C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\Common7\IDE\CommonExtensions\Microsoft\TeamFoundation\Team Explorer;C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Current\bin\Roslyn;C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\Team Tools\Performance Tools\x64;C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\Team Tools\Performance Tools;C:\Program Files (x86)\Microsoft Visual Studio\Shared\Common\VSPerfCollectionTools\vs2019\\x64;C:\Program Files (x86)\Microsoft Visual Studio\Shared\Common\VSPerfCollectionTools\vs2019\;C:\Program Files (x86)\Microsoft SDKs\Windows\v10.0A\bin\NETFX 4.8 Tools\x64\;C:\Program Files (x86)\Windows Kits\10\bin\10.0.18362.0\x64;C:\Program Files (x86)\Windows Kits\10\bin\x64;C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\\MSBuild\Current\Bin;C:\Windows\Microsoft.NET\Framework64\v4.0.30319;C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\Common7\IDE\;C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\Common7\Tools\;;C:\Windows\system32
    SET PWD=/proc/self/cwd
    SET PYTHON_BIN_PATH=C:/Users/mikhail.startsev/Venvs/TF2Build/Scripts/python.exe
    SET PYTHON_LIB_PATH=C:/Users/mikhail.startsev/Venvs/TF2Build/lib/site-packages
    SET RUNFILES_MANIFEST_ONLY=1
    SET TEMP=C:\Users\MIKHAI~1.STA\AppData\Local\Temp
    SET TF2_BEHAVIOR=1
    SET TF_CONFIGURE_IOS=0
    SET TF_CUDA_COMPUTE_CAPABILITIES=3.5,7.0
    SET TF_ENABLE_XLA=1
    SET TF_NEED_CUDA=1
    SET TMP=C:\Users\MIKHAI~1.STA\AppData\Local\Temp
  C:/Users/mikhail.startsev/Venvs/TF2Build/Scripts/python.exe -B external/local_config_cuda/crosstool/windows/msvc_wrapper_for_nvcc.py /nologo /DCOMPILER_MSVC /DNOMINMAX /D_WIN32_WINNT=0x0600 /D_CRT_SECURE_NO_DEPRECATE /D_CRT_SECURE_NO_WARNINGS /D_SILENCE_STDEXT_HASH_DEPRECATION_WARNINGS /bigobj /Zm500 /J /Gy /GF /EHsc /wd4351 /wd4291 /wd4250 /wd4996 /I. /Ibazel-out/x64_windows-opt/bin /Iexternal/eigen_archive /Ibazel-out/x64_windows-opt/bin/external/eigen_archive /Iexternal/local_config_sycl /Ibazel-out/x64_windows-opt/bin/external/local_config_sycl /Iexternal/com_google_absl /Ibazel-out/x64_windows-opt/bin/external/com_google_absl /Iexternal/nsync /Ibazel-out/x64_windows-opt/bin/external/nsync /Iexternal/gif /Ibazel-out/x64_windows-opt/bin/external/gif /Iexternal/libjpeg_turbo /Ibazel-out/x64_windows-opt/bin/external/libjpeg_turbo /Iexternal/com_google_protobuf /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf /Iexternal/com_googlesource_code_re2 /Ibazel-out/x64_windows-opt/bin/external/com_googlesource_code_re2 /Iexternal/farmhash_archive /Ibazel-out/x64_windows-opt/bin/external/farmhash_archive /Iexternal/fft2d /Ibazel-out/x64_windows-opt/bin/external/fft2d /Iexternal/highwayhash /Ibazel-out/x64_windows-opt/bin/external/highwayhash /Iexternal/zlib /Ibazel-out/x64_windows-opt/bin/external/zlib /Iexternal/double_conversion /Ibazel-out/x64_windows-opt/bin/external/double_conversion /Iexternal/snappy /Ibazel-out/x64_windows-opt/bin/external/snappy /Iexternal/local_config_cuda /Ibazel-out/x64_windows-opt/bin/external/local_config_cuda /Iexternal/local_config_tensorrt /Ibazel-out/x64_windows-opt/bin/external/local_config_tensorrt /Iexternal/mkl_dnn /Ibazel-out/x64_windows-opt/bin/external/mkl_dnn /Iexternal/curl /Ibazel-out/x64_windows-opt/bin/external/curl /Iexternal/boringssl /Ibazel-out/x64_windows-opt/bin/external/boringssl /Iexternal/jsoncpp_git /Ibazel-out/x64_windows-opt/bin/external/jsoncpp_git /Iexternal/aws /Ibazel-out/x64_windows-opt/bin/external/aws /Iexternal/aws-c-common /Ibazel-out/x64_windows-opt/bin/external/aws-c-common /Iexternal/aws-c-event-stream /Ibazel-out/x64_windows-opt/bin/external/aws-c-event-stream /Iexternal/aws-checksums /Ibazel-out/x64_windows-opt/bin/external/aws-checksums /Iexternal/com_github_grpc_grpc /Ibazel-out/x64_windows-opt/bin/external/com_github_grpc_grpc /Iexternal/upb /Ibazel-out/x64_windows-opt/bin/external/upb /Ibazel-out/x64_windows-opt/bin/external/local_config_cuda/cuda/_virtual_includes/cuda_headers_virtual /Ibazel-out/x64_windows-opt/bin/external/local_config_tensorrt/_virtual_includes/tensorrt_headers /Ibazel-out/x64_windows-opt/bin/external/local_config_cuda/cuda/_virtual_includes/cublas_headers_virtual /Ibazel-out/x64_windows-opt/bin/external/local_config_cuda/cuda/_virtual_includes/cudnn_header /Iexternal/eigen_archive /Ibazel-out/x64_windows-opt/bin/external/eigen_archive /Iexternal/nsync/public /Ibazel-out/x64_windows-opt/bin/external/nsync/public /Iexternal/gif /Ibazel-out/x64_windows-opt/bin/external/gif /Iexternal/gif/windows /Ibazel-out/x64_windows-opt/bin/external/gif/windows /Iexternal/com_google_protobuf/src /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src /Iexternal/farmhash_archive/src /Ibazel-out/x64_windows-opt/bin/external/farmhash_archive/src /Iexternal/zlib /Ibazel-out/x64_windows-opt/bin/external/zlib /Iexternal/double_conversion /Ibazel-out/x64_windows-opt/bin/external/double_conversion /Iexternal/local_config_cuda/cuda /Ibazel-out/x64_windows-opt/bin/external/local_config_cuda/cuda /Iexternal/local_config_cuda/cuda/cuda/include /Ibazel-out/x64_windows-opt/bin/external/local_config_cuda/cuda/cuda/include /Iexternal/local_config_cuda/cuda/cublas/include /Ibazel-out/x64_windows-opt/bin/external/local_config_cuda/cuda/cublas/include /Iexternal/mkl_dnn/include /Ibazel-out/x64_windows-opt/bin/external/mkl_dnn/include /Iexternal/mkl_dnn/src /Ibazel-out/x64_windows-opt/bin/external/mkl_dnn/src /Iexternal/mkl_dnn/src/common /Ibazel-out/x64_windows-opt/bin/external/mkl_dnn/src/common /Iexternal/mkl_dnn/src/cpu /Ibazel-out/x64_windows-opt/bin/external/mkl_dnn/src/cpu /Iexternal/mkl_dnn/src/cpu/gemm /Ibazel-out/x64_windows-opt/bin/external/mkl_dnn/src/cpu/gemm /Iexternal/mkl_dnn/src/cpu/xbyak /Ibazel-out/x64_windows-opt/bin/external/mkl_dnn/src/cpu/xbyak /Iexternal/curl/include /Ibazel-out/x64_windows-opt/bin/external/curl/include /Iexternal/boringssl/src/include /Ibazel-out/x64_windows-opt/bin/external/boringssl/src/include /Iexternal/jsoncpp_git/include /Ibazel-out/x64_windows-opt/bin/external/jsoncpp_git/include /Iexternal/aws/aws-cpp-sdk-core/include /Ibazel-out/x64_windows-opt/bin/external/aws/aws-cpp-sdk-core/include /Iexternal/aws/aws-cpp-sdk-s3/include /Ibazel-out/x64_windows-opt/bin/external/aws/aws-cpp-sdk-s3/include /Iexternal/aws/aws-cpp-sdk-transfer/include /Ibazel-out/x64_windows-opt/bin/external/aws/aws-cpp-sdk-transfer/include /Iexternal/aws-c-common/include /Ibazel-out/x64_windows-opt/bin/external/aws-c-common/include /Iexternal/aws-c-event-stream/include /Ibazel-out/x64_windows-opt/bin/external/aws-c-event-stream/include /Iexternal/aws-checksums/include /Ibazel-out/x64_windows-opt/bin/external/aws-checksums/include /Iexternal/com_github_grpc_grpc/include /Ibazel-out/x64_windows-opt/bin/external/com_github_grpc_grpc/include /Iexternal/com_github_grpc_grpc/src/core/ext/upb-generated /Ibazel-out/x64_windows-opt/bin/external/com_github_grpc_grpc/src/core/ext/upb-generated /Iexternal/com_github_grpc_grpc/third_party/address_sorting/include /Ibazel-out/x64_windows-opt/bin/external/com_github_grpc_grpc/third_party/address_sorting/include /DEIGEN_MPL2_ONLY /DEIGEN_MAX_ALIGN_BYTES=64 /DEIGEN_HAS_TYPE_TRAITS=0 /D__CLANG_SUPPORT_DYN_ANNOTATION__ /DTF_USE_SNAPPY /DTENSORFLOW_USE_CUSTOM_CONTRACTION_KERNEL /DTENSORFLOW_USE_MKLDNN_CONTRACTION_KERNEL /DCURL_STATICLIB /DPLATFORM_WINDOWS /DENABLE_CURL_CLIENT /DOPENSSL_IS_BORINGSSL /DGRPC_ARES=0 /showIncludes /MD /O2 /DNDEBUG /w /D_USE_MATH_DEFINES -DWIN32_LEAN_AND_MEAN -DNOGDI /std:c++14 -x cuda -DGOOGLE_CUDA=1 --no-cuda-include-ptx=all --cuda-include-ptx=sm_35 --cuda-gpu-arch=sm_35 --cuda-include-ptx=sm_70 --cuda-gpu-arch=sm_70 -DGOOGLE_CUDA=1 -DTENSORFLOW_USE_NVCC=1 -DTENSORFLOW_USE_XLA=1 -DTENSORFLOW_MONOLITHIC_BUILD /DPLATFORM_WINDOWS /DEIGEN_HAS_C99_MATH /DTENSORFLOW_USE_EIGEN_THREADPOOL /DEIGEN_AVOID_STL_ARRAY /Iexternal/gemmlowp /wd4018 /wd4577 /DNOGDI /DTF_COMPILE_LIBRARY -nvcc_options=relaxed-constexpr -nvcc_options=ftz=true /Fobazel-out/x64_windows-opt/bin/tensorflow/core/kernels/_objs/debug_ops_gpu/debug_ops_gpu.cu.o /c tensorflow/core/kernels/debug_ops_gpu.cu.cc
ERROR: C:/source/tensorflow/tensorflow/core/kernels/BUILD:1013:1: C++ compilation of rule '//tensorflow/core/kernels:debug_ops_gpu' failed (Exit 255)
**( was unexpected at this time.**
Target //tensorflow/core/kernels:debug_ops_gpu failed to build
INFO: Elapsed time: 1110.066s, Critical Path: 114.51s
INFO: 2748 processes: 2748 local.
FAILED: Build did NOT complete successfully
```

It is not the error of the msvc_wrapper_for_nvcc.py, since running it separately (with all the commands listed here) did not result in an error.

It seems that this type of errors ("XXX was unexpected at this time") are cmd-specific, but the generic solution of `setlocal enableextensions enabledelayedexpansion` did not help or anyhow change anything.

**Other attempts with the same results**
* Also tried [building from the MSYS shell](https://www.tensorflow.org/install/source_windows#build_using_the_msys_shell)
* Tried using cygwin sh as BAZEL_SH
* TF2.1.0 (as well as TF2.1.1) and the master branch from the TF repo, with respective Bazel versions - 0.29.1 and 3.0.0

What _did_ work was building the tensorflow.dll without GPU support requested in configure.cmd, but that is not what I wanted to have in the end...

Overall, it seems that some batch command/script generated/used during building is malformatted, maybe because some environment variable is not set (and is expanded into and empty string) or something else, but I don't have a way to know which one, or which command exactly fails. The `-s` option to `bazel build` does not output the precise command that results into the error I am experiencing (and setting --logging to the highest value does not help either).

**Comment by gunan** [comment]
@Artem-B @chsigg any ideas?

**Comment by chsigg** [comment]
> running it separately (with all the commands listed here) did not result in an error.
I think the message is from bazel when the 'compiler' (msvc_wrapper_for_nvcc.py) returns non-zero.
Would you mind checking the result code of running the command explicitly? And maybe add some manual logging to the wrapper to figure out what's going on?
Sorry I don't have any better advice. I just had a similar issue last week, and couldn't get the --copt=--cuda_log to work either and reverted to printf debugging. :-(

**Comment by Artem-B** [comment]
No idea what's going on, but it should be possible to narrow the issues down a bit.
* add printouts in the `msvc_wrapper_for_nvcc.py` to check whether we actually get to launch nvcc.
* if we don't get to the wrapper, the issues is somewhere in bazel
* if the wrapper launches, but we fail to launch nvcc, then there's some sort of issue with python.
* if we do launch nvcc, modify the wrapper to pass `-v` and see whether the failure happens there and if so, at which point.

**Comment by MikhailStartsev** [comment]
@chsigg @Artem-B Thank you for the suggestions for debugging! Following your recommendations:
* It does get to the msvc_wrapper_for_nvcc.py call, and this finishes successfully (%errorlevel% is 0)
* Because of this, adding -v to the wrapper (as well as calling it with --cuda_log flag) does not shed any extra light on the fail reason...

To me, the symptoms look similar to something like this: https://stackoverflow.com/a/10511644/1675506 , but, as I mentioned, `setlocal enableextensions enabledelayedexpansion` does not help in this case... Is there maybe a way to log all cmd command executed by bazel?.. I did not find one, at least. Or a list of _all_ environment variables that are required for this build (for GPU-related targets specifically) would be nice, but I don't see it documented either..

**Comment by gunan** [comment]
You can use the `-s` flag of bazel to log all commands invoked by bazel.

**Comment by MikhailStartsev** [comment]
@gunan I used it in the initial post already, hoping to catch the actually failing command, but it is not listed (see the output log I posted), unfortunately - the command sequence after which the crash happens completes successfully.

**Another update:** I set all the variables that were used in the CI build log but that I did not set before (i.e. GCS_BUCKET_NAME, GOOGLE_CLOUD_CREDENTIAL, BAZEL_WRAPPER_PATH, PYTHON_DIRECTORY, BUILD_PATH, GEN_SCRIPT, and GEN_BUILD) to the strings with the same content to at least maybe get a different error, but to no avail.

**Comment by Artem-B** [comment]
I'm fresh out of ideas. Debugging windows issues is such a pain. I can't wait until WSL2 supports GPUs so we could just use linux build. @gunan -- do we have any contacts at MSFT working on/with TF? There's a good chance that someone out there does have the know-how of dealing with these kinds of issues.

**Comment by jgasperlin** [comment]
I build tag v2.2.0 with command 
`bazel build --config=opt --config=cuda --define=no_tensorflow_py_deps=true //tensorflow/tools/lib_package:libtensorflow`

**Comment by jgasperlin** [comment]
Make sure you install build tools to C:\msys64, since they don't digest space in 'Program Files'.

**Comment by MikhailStartsev** [comment]
@jgasperlin I tried installing VS Build Tools separately (was using my existing VS installation before; installed into a msys64 subfolder as you suggested) - get exactly the same error as before. Don't think our issues are related, to be honest...

**Comment by jgasperlin** [comment]
Ok we misunderstood as build tool i meant `Msys` that resides on `C:\msys64` by default  path (Just checked)
BAZEL_VS where you set your build tool (Mine is 2019 ) can have spaces and equals to 
`C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC`
I remembered i had similar problem and if not Msys path i think it was python then mine is on
`where python` => `C:\Users\Targma\AppData\Local\Programs\Python\Python38\python.exe`
which is probably same for you but then i switched Bazel 3.2 for 2.0 which build status is confirmed on 
https://www.tensorflow.org/install/source_windows
if you scroll to the bottom.
>Version | Python version | Compiler | Build tools | cuDNN | CUDA
tensorflow_gpu-2.2.0 | 3.5-3.8 | MSVC 2019 | Bazel 2.0.0 | 7.4 | 10.1

I just resolved also issues with symbols which you should check out https://github.com/tensorflow/tensorflow/issues/39407

Hope that helps, you should also run Bazel with verbose logging -s (https://stackoverflow.com/questions/32823625/bazel-build-verbose-compiler-commands-logging) so you know where the problem is

**Comment by MikhailStartsev** [comment]
@jgasperlin Ah, okay, I see. Either way, my msys64 path has no spaces in it as it is, yes. And the python path is without spaces as well... I am using the recommended bazel version from the same list you are referring to. The only difference is that cuDNN 7.4 should not support cuda 10.1, so I am using a later cuDNN version... I also used the `-s` flag, yes, but it does not print the command that _actually_ results in the error I get...

Looking at the issue you linked, I tried compiling a different target (//tensorflow/tools/lib_package:libtensorflow) instead of the one I used, with the same error coming out in the end... (i.e. "( was unexpected at this time.")

I don't get as far as the linker errors you were experiencing, but thanks for the heads-up, in case it gets to it!

Since you succeeded in building tensorflow.dll with CUDA support under Windows, could you please post the exact command sequence you use for this (including setting the PATH and other necessary variables - from a "clean" terminal, without anything extra pre-set) - I am afraid that I have some environment variable missing, but can't figure out which one...

Thanks for responding so far :) It is encouraging that you managed to make it work!

**Comment by jgasperlin** [comment]
make sure before you rebuild anything with Bazel do 
`bazel clean --expunge` this will clean up everything
i used lates cuDNN 7.6.something
do you have set cuda path ? when you configure with 
`python configure.py`
does it pick Cuda v10.1 and cuDNN 7.6 ?
My problem was resolved with using correct bazel to build (2.0.0)

**Comment by Dave-Stein** [comment]
@jgasperlin  
I have been asking something like this since 6 months ago. No progress for anything related to tensorflow2.*

https://storage.googleapis.com/tensorflow/libtensorflow/libtensorflow-gpu-windows-x86_64-2.0.0.zip


Any chance if you can provide a link to your gpu version of libtensorflow?  So we can continue to test while Google finally makes up the mind to support Windows developers

**Comment by Dave-Stein** [comment]
I think the only way forwards is to start a campaign that WDM (Windows Developers Matter)  :-)

**Comment by Dave-Stein** [comment]
We need to include this =>  label:"stat:awaiting response"

**Comment by gunan** [comment]
@av8ramit @bmzhao is working to fix libtensorflow releases.
As mentioned in other issues, owners of our C++ APIs have left, and now we are ramping up on the releases, moving them to 2.0, and cleaning up stale build scripts all at the same time. This is taking some time, thus the delay. Moreover, we have lost multiple people with windows expertise, so that has been quite difficult for us to ramp up as well.

**Comment by Dave-Stein** [comment]
@gunan  => @av8ramit @bmzhao  were never assigned BUT U to these issues last few months.

**Comment by Dave-Stein** [comment]
@av8ramit  support LINUX last one month, not WINDOWS!!

**Comment by Dave-Stein** [comment]
@bmzhao has no records of contributing to WINDOWS! last 3 months

**Comment by Dave-Stein** [comment]
We need to include this => label:"stat:awaiting response"

**Comment by Dave-Stein** [comment]
We need to invite more discussion on the future of WINDOW support for tensorflow.  6 months with no progress but challenges. **Google needs to hire those WHO CLIAM HERE are able to address these challenges IMMEDIATELY.**

**Comment by Dave-Stein** [comment]
The latest commit to address this problem is to get more Windows developers to spend MORE TIME trying [new Bazel  (from 2.0 to 3.0)](https://github.com/tensorflow/tensorflow/commit/b4b83222d470afbf0b83d12b0824c0f056235655)

**Comment by Dave-Stein** [comment]
Next step is to do DATA SCIENCE, how many TOTAL MONTHS, all windows developers HERE spent on (Attempting) building tensorlfow GPU DLL!

**Comment by Dave-Stein** [comment]
@jgasperlin  Google should hire you to avoid the WINDOWS developers spending TOTAL many months just to build Windows dll GPU

**Comment by Dave-Stein** [comment]
@gunan  NEXT step is to request/invite discussions at Linkedin on this issue as this is urgent and NO progress has been made last 6 months

**Comment by Dave-Stein** [comment]
> @av8ramit @bmzhao is working to fix libtensorflow releases.
> As mentioned in other issues, owners of our C++ APIs have left, and now we are ramping up on the releases, moving them to 2.0, and cleaning up stale build scripts all at the same time. This is taking some time, thus the delay. Moreover, we have lost multiple people with windows expertise, so that has been quite difficult for us to ramp up as well.

@gunan  NEXT step is to request/invite discussions at Linkedin on this issue as this is urgent and NO progress has been made last 6 months

**Comment by ewilderj** [comment]
I have blocked @Dave-Stein for now because of the spamming behavior. Please get in touch Dave if you wish to be unblocked and abide within our community rules for civil and constructive discussion.

**Comment by MikhailStartsev** [comment]
Ooookay, finally an update! @gunan @Artem-B @av8ramit @bmzhao, in case you are interested, the issue that I was experiencing before was connected to the fact that bazel **clears the environment** at some point or other during its build ahead of some operations. Turned out that my system had a special env variable set up, which led to the error I saw being produced by a script that is executed when any new cmd is created on my system. 

Therefore, so far: (1) this is not a general error that will affect everyone, but (2) I think it should be made clear somewhere in the documentation that certain commands during the build will be executed in a cleaned environment, so one cannot rely on the original env variables being available throughout the build, even if they are not used or set by the configuration script or bazel directly. **This can cause all sorts of issues and artefacts on different systems, depending on the set-up**, would be at least good to know this upfront, if not possibly have a workaround for this (to keep certain or all variables intact).

I still have some issues, the latest being "Failed to run \c\Users\mikhail.startsev\AppData\Local\Temp/tmpxft_00004654_00000000-1.bat (The system cannot find the file specified.", but at least it's a start :)

**Comment by MikhailStartsev** [comment]
Okay, closing the issue, it works now - having initially had a suspicion that it has to do with env variables, just not knowing which ones, I messed with the environement a lot, this seems to have cause some temporary issue. Now building with my "normal" env + setting bazel-specific variables works fine!

Thanks everyone for your responses. Just so that this does not fall through the cracks, repeating again: I still think this behaviour should be explicitly documented.

The solution for preserving some env variable that I got recommended on the bazel mailing list is [--action_env](https://docs.bazel.build/versions/master/command-line-reference.html#flag--action_env).
